Added null check to some adapters and changed type definition

ayame113 opened the pull request:
Some LSP requests lack null checking and are causing errors, so I fixed them.

![image](https://user-images.githubusercontent.com/40050810/115957991-5ce60300-a540-11eb-945e-057e1b28ae36.png)

I made the following changes:
1. Added null check (https://github.com/atom-community/atom-languageclient/commit/e20bba5a4fe06b4f1a24f2063b1a60fcd5da1514)
2. Fixed the type definition (https://github.com/atom-community/atom-languageclient/commit/85cc505a5c9999e4da044996aa49e436d32d017a)
3. Use the vscode-languageserver-protocol package for type definition (https://github.com/atom-community/atom-languageclient/compare/85cc505..d665159)

If you don't need the third change, reverting it will still work.

UziTech commented:
I think we were discussing this in https://github.com/atom-community/atom-languageclient/pull/132#discussion_r595580845

I'm not sure if it is better to through an error if the lsp returns something invalid or fail silently like this PR. Is null or undefined a valid return for `.codeAction`?

aminya commented:
> I think we were discussing this in [#132 (comment)](https://github.com/atom-community/atom-languageclient/pull/132#discussion_r595580845)

This is different though.

UziTech commented:
> This is different though.

How? Why are these checks necessary if `null` or `undefined` are not correct outputs of `.codeAction` anyway?

aminya commented:
I meant the changes under `lib/languageclient.ts` which seem reasonable as they are using the original types.

UziTech commented:
> I meant the changes under `lib/languageclient.ts` which seem reasonable as they are using the original types.

Ya I think those changes are good. I'm just not sure about null checks.

ayame113 commented:
(some comment)
- Currently, the [`vscode-languageserver-protocol` module](https://www.npmjs.com/package/vscode-languageserver-protocol) is used for type definition. In this PR, I also want to use it in the return value of the language-client method.
 - Since the return type is not exported directly from the module, I am using the generics of the function.
   - I used it as a reference for implementation
https://github.com/microsoft/vscode-languageserver-node/blob/d58c00bbf8837b9fd0144924db5e7b1c543d839e/client/src/common/client.ts#L83-L86
 - This will cause a type error if the wrong return type.
    ![image](https://user-images.githubusercontent.com/40050810/115989078-78f8ab80-a5f7-11eb-95e5-ebedbfd48848.png)
- Removed null from the return value of the following request according to the type definition
  - completionItem / resolve
  - codeLens / resolve
 - Added null
   - textDocument/references
   - textDocument/documentHighlight
   - textDocument/documentSymbol
   - workspace/symbol
   - textDocument/codeAction
   - textDocument/codeLens
   - textDocument/documentLink
   - textDocument/formatting
   - textDocument/rangeFormatting
   - textDocument/onTypeFormatting
   - textDocument/rename
 - I think this implementation fits the spec, but you can leave null in the return value if you want to deal with invalid null checks outside the spec.

UziTech commented on lib/adapters/code-action-adapter.ts:
```diff
@@ -47,7 +47,7 @@ export default class CodeActionAdapter {
 
     const params = CodeActionAdapter.createCodeActionParams(linterAdapter, editor, range, diagnostics)
     const actions = await connection.codeAction(params)
-    return actions.map((action) => CodeActionAdapter.createCodeAction(action, connection))
+    return (actions || []).map((action) => CodeActionAdapter.createCodeAction(action, connection))
```
According to [`codeAction`](https://github.com/atom-community/atom-languageclient/blob/ffbe675bf8a2dc90171a5599dd2dcf88a4398f0f/lib/languageclient.ts#L403) it returns `Promise<Array<lsp.Command | lsp.CodeAction>>` so a promise resolving to `null` or `undefined` is an invalid return type and should probably throw an error instead of being handled the same as an empty array. Or the return type should be updated to include `null` and `undefined` as valid returns.

UziTech replied:
I tend to prefer throwing a valid error instead of failing silently like this. @aminya This is what I was talking about in https://github.com/atom-community/atom-languageclient/pull/132#discussion_r595452569 when I said "types do not catch runtime errors".

ayame113 replied:
Thank you for your review!

The spec seems to say that codeAction returns `(Command | CodeAction) [] | null`. Is it not necessary to raise an error if the language server returns valid null?
> https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_codeAction
> Request:
> - method: ‘textDocument/codeAction’
> 
> Response:
> - result: (Command | CodeAction)[] | null

By the way, such a change is also an option.
```diff
 public static async getCodeActions(
   ...
-): Promise<atomIde.CodeAction[]> {
+): Promise<atomIde.CodeAction[] | null> {
   if (linterAdapter == null) {
     return []
   }
   assert(serverCapabilities.codeActionProvider, "Must have the textDocument/codeAction capability")
   const params = CodeActionAdapter.createCodeActionParams(linterAdapter, editor, range, diagnostics)
   const actions = await connection.codeAction(params)
-  return (actions || []).map((action) => CodeActionAdapter.createCodeAction(action, connection))
+  if (actions==null) {return null}
+  return actions.map((action) => CodeActionAdapter.createCodeAction(action, connection))
 }
```

UziTech replied:
I like the early null check rather than still mapping over an empty array.

We should change the return type of [codeAction](https://github.com/atom-community/atom-languageclient/blob/ffbe675bf8a2dc90171a5599dd2dcf88a4398f0f/lib/languageclient.ts#L403) if null is a valid return according to the spec.

ayame113 replied:
At https://github.com/atom-community/atom-languageclient/pull/150/commits/e7c26b4f22dbf58ae17788b2a66fa04abeafc466, I changed it to return null instead of an empty array when `actions` value is null.

UziTech commented on lib/languageclient.ts:
```diff
@@ -421,8 +421,8 @@ export class LanguageClientConnection extends EventEmitter {
    */
-  public codeLensResolve(params: lsp.CodeLens): Promise<lsp.CodeLens | null> {
-    return this._sendRequest("codeLens/resolve", params)
+  public codeLensResolve(params: lsp.CodeLens): Promise<lsp.CodeLens> {
+    return this._sendRequest(lsp.CodeLensResolveRequest.type, params)
```
Would this be a breaking change since we are changing a public API?

ayame113 replied:
The LanguageClient class has not been exported from main.ts, but it may still be a breaking change. I do not know well which.😰

UziTech replied:
Maybe just a note in the changelogs

aminya commented:
I'll review today or tomorrow
